crates/sui-move-build/src/linters/freeze_wrapped.rs

```diff
@@ -0,0 +1,418 @@
+    wrapped_type_loc: Loc,
+    /// Is the wrapping direct or indirect
+    direct: bool,
+}
+
+/// Structs (per-module) that have fields with the key ability
+type KeyFields = BTreeMap<E::ModuleIdent, BTreeMap<P::StructName, Option<KeyFieldInfo>>>;
+
+#[derive(Default)]
+pub struct FreezeWrappedVisitor {
+    /// Memoizes information about struct fields with key ability as they are discovered
+    key_fields: KeyFields,
+}
+
+impl TypingVisitor for FreezeWrappedVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &ProgramInfo,
+        program: &mut T::Program,
+    ) {
+        for (_, _, mdef) in program.modules.iter() {
+            env.add_warning_filter_scope(mdef.warning_filter.clone());
+
+            for (_, _, fdef) in mdef.functions.iter() {
+                env.add_warning_filter_scope(fdef.warning_filter.clone());
+
+                if let T::FunctionBody_::Defined(seq) = &fdef.body.value {
+                    self.visit_seq(seq, env, &program.modules);
+                }
+
+                env.pop_warning_filter_scope();
+            }
+
+            env.pop_warning_filter_scope();
+        }
+    }
+}
+
+impl FreezeWrappedVisitor {
+    fn visit_seq_item(
+        &mut self,
+        sp!(_, seq_item): &T::SequenceItem,
+        env: &mut CompilationEnv,
+        modules: &UniqueMap<E::ModuleIdent, T::ModuleDefinition>,
+    ) {
+        use T::SequenceItem_ as SI;
+        match seq_item {
+            SI::Seq(e) => self.visit_exp(e, env, modules),
+            SI::Declare(_) => (),
+            SI::Bind(_, _, e) => self.visit_exp(e, env, modules),
+        }
+    }
+
+    fn visit_exp(
+        &mut self,
+        exp: &T::Exp,
+        env: &mut CompilationEnv,
+        modules: &UniqueMap<E::ModuleIdent, T::ModuleDefinition>,
+    ) {
+        use T::UnannotatedExp_ as E;
+        let sp!(_, uexp) = &exp.exp;
+        match uexp {
+            E::ModuleCall(fun) => {
+                if FREEZE_FUNCTIONS.iter().any(|(addr, module, fname)| {
+                    fun.module.value.is(*addr, *module) && &fun.name.value().as_str() == fname
+                }) {
+                    let Some(bt) = base_type(&fun.type_arguments[0]) else {
+                        // not an (potentially dereferenced) N::Type_::Apply nor N::Type_::Param
+                        return;
+                    };
+                    let N::Type_::Apply(_,tname, _) = &bt.value else {
+                        // not a struct type
+                        return;
+                    };
+                    let N::TypeName_::ModuleType(mident, sname) = tname.value else {
+                        // struct with a given name not found
+                        return;
+                    };
+                    // see if info about fields of a give struct was already found
+                    let (key_field_info, info_inserted) = self.get_key_field(&mident, &sname);
```

Comment:
It's fine, but I feel like I would try to encapsulate all of this logic inside of `struct_fields` somehow? Just feels weird that this area of the visitor knows about the memoization elsewhere

Reply:
It was a little more complicated than I thought, but I think it does look better now. Thank you for the suggestion! I also did some renaming (from "key_field" to "wrapping_field" which I think is more intuitive, especially considering code comment explanations that were already there). Please take another look if you'd like!